### PR TITLE
Server + token authentication + simple routes

### PR DESCRIPTION
**Pull request by mateuszmatyszkowicz**

### CODE REVIEW

File: app/controllers/user/delete-all.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+const { User } = require('../../models');
+
+module.exports = (req, res, next) => {
+    User.remove({})
+        .exec()
+        .then((result) => {
+            console.log(result);
```
Comment (kamilogerto2) [inline app/controllers/user/delete-all.js]: I think that you can use some Logger - avoid console.log.

File: app/controllers/user/delete-all.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+            if (result.n) {
+                return res.status(200).json({
```
Comment (kamilogerto2) [inline app/controllers/user/delete-all.js]: Please use https://www.npmjs.com/package/http-status-codes for statuses.

File: app/controllers/user/delete-all.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+                    message: 'Users deleted',
```
Comment (kamilogerto2) [inline app/controllers/user/delete-all.js]: We don't need this information - status should be sufficient.

File: app/controllers/user/delete-all.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+                });
+            }
+
+            res.status(409).json({
+                message: 'Errors occured.',
```
Comment (kamilogerto2) [inline app/controllers/user/delete-all.js]: We don't need this information - status should be sufficient.

File: app/controllers/user/delete.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+const { User } = require('../../models');
+
+module.exports = (req, res, next) => {
+    User.remove({ _id: req.params.id})
+        .exec()
+        .then((result) => {
+            console.log(result);
+            if (result.n) {
+                return res.status(200).json({
+                    message: 'User deleted',
```
Comment (kamilogerto2) [inline app/controllers/user/delete.js]: We don't need this information - status should be sufficient.

File: app/controllers/user/delete.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+                });
+            }
+
+            res.status(409).json({
+                message: 'User not found',
```
Comment (kamilogerto2) [inline app/controllers/user/delete.js]: We don't need this information - status should be sufficient.

File: app/controllers/user/delete.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
+const { User } = require('../../models');
+
+module.exports = (req, res, next) => {
+    User.remove({ _id: req.params.id})
+        .exec()
+        .then((result) => {
+            console.log(result);
+            if (result.n) {
```
Comment (kamilogerto2) [inline app/controllers/user/delete.js]: what is result.n

File: app/controllers/user/get-all.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    User.find({})
+        .exec()
+        .then((users) => {
+            if (users) {
+                return res.status(200).json({
+                    users: users,
+                });
+            }
+
+            res.status(404).json({
```
Comment (kamilogerto2) [inline app/controllers/user/get-all.js]: Why we can't return an empty array?

File: app/controllers/user/login.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+const JWT_KEY = require('../../../config/config').jwt.secret;
+
+module.exports = (req, res, next) => {
+    User.find({ 'local.email': req.body.email })
+        .exec()
+        .then((users) => {
+            console.log(`Trying to log in as ${users[0].local.email}`);
+            if (users.length < 1) {
+                return res.status(404).json({
+                    message: 'Please provide correct email address',
```
Comment (kamilogerto2) [inline app/controllers/user/login.js]: We don't need this information - status should be sufficient.

Comment (kamilogerto2) [reply app/controllers/user/login.js]: add status unauhtorized

Comment (kamilogerto2) [reply app/controllers/user/login.js]: We can't tell user if email or password is bad.

File: server.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+const bodyParser = require('body-parser');
+const cookieParser = require('cookie-parser');
+const flash = require('connect-flash');
+
+mongoose.connect('mongodb://localhost:27017/coders');
+
+app.use(morgan('dev'));
+app.use(bodyParser.json());
+    app.use(bodyParser.urlencoded({ extended: true }));
```
Comment (kamilogerto2) [inline server.js]: delete tab

File: package.json [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    "bcrypt": "^1.0.3",
+    "body-parser": "^1.18.2",
+    "connect-flash": "^0.1.1",
+    "cookie-parser": "^1.4.3",
+    "express": "^4.16.2",
+    "express-session": "^1.15.6",
+    "jwt-simple": "^0.5.1",
+    "mongoose": "^5.0.1",
+    "morgan": "^1.9.0",
+    "passport": "^0.4.0",
```
Comment (kamilogerto2) [inline package.json]: We use passport?

File: app/models/user.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        bcrypt.compare(passwordCandidate, password, (err, success) => {
+            if (err) {
+                return reject(err);
+            }
+
+            if (success) {
+                return resolve(success);
+            }
+
+            return reject();
```
Comment (kamilogerto2) [inline app/models/user.js]: why return?

File: app/models/user.js [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    })
+});
+
+userSchema.methods.comparePassword = function(passwordCandidate) {
+    let password = this.local.password;
+
+    return new Promise((resolve, reject) => {
+        bcrypt.compare(passwordCandidate, password, (err, success) => {
+            if (err) {
+                return reject(err);
```
Comment (kamilogerto2) [inline app/models/user.js]: why return?

File: app/models/user.js [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+            }
+
+            if (success) {
+                return resolve(success);
```
Comment (kamilogerto2) [inline app/models/user.js]: why return?

File: app/models/user.js [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+            type: String,
+            required: true,
+        },
+    },
+});
+
+userSchema.pre('save', function (next) {
+    bcrypt.hash(this.local.password, 10, (err, hash) => {
+        if (err) {
+            return next(err);
```
Comment (kamilogerto2) [inline app/models/user.js]: why return?

**Review by kamilogerto2 — CHANGES_REQUESTED** [review]

File: app/controllers/user/delete.js (hunk 7 above)
Comment (mateuszmatyszkowicz) [reply app/controllers/user/delete.js]: `result.n` is number of successfully removed documents.

File: server.js [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+const mongoose = require('mongoose');
+const morgan = require('morgan');
+const bodyParser = require('body-parser');
+const cookieParser = require('cookie-parser');
+
+if (!fs.existsSync(config.logger.path)) {
+    fs.mkdirSync(config.logger.path);
+};
+
+mongoose.connect('mongodb://localhost:27017/coders');
```
Comment (kamilogerto2) [inline server.js]: address could be in config

File: server.js [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+
+const logger = require('./src/lib/logger');
```
Comment (kamilogerto2) [inline server.js]: move const to other consts

File: src/controllers/auth/login.js [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+
+const { User } = require('../../models');
+
+module.exports = (req, res, next) => {
+    User.findOne({ 'local.email': req.body.email })
+        .exec()
+        .then((user) => {
+            logger.info(`TRY$ LOGIN ${req.body.email}`);
+            if (!user) {
+                return res.sendStatus(HttpStatus.NOT_FOUND);
```
Comment (kamilogerto2) [inline src/controllers/auth/login.js]: please return here also status UNAUTHORIZED

File: src/controllers/auth/login.js [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+                        userId: user._id,
+                    },
+                        JWT_SECRET
+                    , {
+                        expiresIn: JWT_EXPIRATION,
+                    });
+
+                    logger.verbose(`USER#NEW_TOKEN ${token}`);
+                    return res.status(HttpStatus.OK).json({
+                        message: 'Auth successful',
```
Comment (kamilogerto2) [inline src/controllers/auth/login.js]: we don't need this message

File: server.js [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+app.use((req, res, next) => {
+    res.header("Access-Control-Allow-Origin", "*");
+    res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
+
+    next();
+});
+
+require('./src/routes')(app);
+
+app.listen(config.server.port, () => {
```
Comment (kamilogerto2) [inline server.js]: unnecessary curly brackets

File: src/controllers/auth/login.js [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+                        token: token,
+                    });
+                })
+                .catch((error) => {
```
Comment (kamilogerto2) [inline src/controllers/auth/login.js]: unnecessary curly brackets

File: src/controllers/user/delete-all.js [hunk 22]
```diff
@@ -0,0 +1,18 @@
+const HttpStatus = require('http-status-codes');
+const logger = require('../../lib/logger');
+
+const { User} = require('../../models');
+
+module.exports = (req, res, next) => {
+    User.remove({})
+        .exec()
+        .then((result) => {
```
Comment (kamilogerto2) [inline src/controllers/user/delete-all.js]: unnecessary curly brackets

File: src/controllers/user/delete.js [hunk 23]
```diff
@@ -0,0 +1,18 @@
+const HttpStauts = require('http-status-codes');
+const logger = require('../../lib/logger');
+
+const { User } = require('../../models');
+
+module.exports = (req, res, next) => {
+    User.remove({ _id: req.params.id})
+        .exec()
+        .then((result) => {
```
Comment (kamilogerto2) [inline src/controllers/user/delete.js]: unnecessary curly brackets

File: src/models/user.js [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+userSchema.methods.comparePassword = function (passwordCandidate) {
+    let password = this.local.password;
+
+    return new Promise((resolve, reject) => {
+        bcrypt.compare(passwordCandidate, password, (err, success) => {
+            if (err) {
+                reject(err);
+            }
+
+            if (success) {
```
Comment (kamilogerto2) [inline src/models/user.js:47]: Is there any case that success will be undefined or null? if not - we don't need if statement

**Review by kamilogerto2 — CHANGES_REQUESTED** [review]

File: src/routes/user/index.js [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+
+
+// Routes
+router.post('/', UserCtrl.signup);
+
+router.delete('/', isAuthenticated, UserCtrl.deleteAll);
+router.delete('/:id', isAuthenticated, UserCtrl.delete);
+
+router.get('/', isAuthenticated, UserCtrl.getAll);
+router.get('/:id', isAuthenticated, UserCtrl.get);
```
Comment (kamilogerto2) [inline src/routes/user/index.js:21]: where is login? :)

File: src/models/user.js (hunk 24 above)
Comment (mateuszmatyszkowicz) [reply src/models/user.js:47]: compare function take a callback function with two parameters, error and results of password comparing, so in this case we can take also false, then we should reject our promise. Maybe name `success` is not propirate

File: src/controllers/user/delete.js [hunk 26]
```diff
@@ -0,0 +1,16 @@
+const HttpStauts = require('http-status-codes');
+const logger = require('../../lib/logger');
+
+const { User } = require('../../models');
+
+module.exports = (req, res, next) => {
+    User.remove({ _id: req.params.id})
```
Comment (kamilogerto2) [inline src/controllers/user/delete.js:7]: add space

File: src/controllers/user/get.js [hunk 27]
```diff
@@ -0,0 +1,8 @@
+const { User } = require('../../models');
+const HttpStatus = require('http-status-codes');
+
+module.exports = (req, res, next) => {
```
Comment (kamilogerto2) [inline src/controllers/user/get.js:4]: unnecessary brackets

**Review by kamilogerto2 — APPROVED** [review]

**Pull request merged**